### PR TITLE
reduction of api calls and acceleration of the inventory building process

### PR DESCRIPTION
**Pull request by richardtief**
Reduce API calls to a minimum to keep the load on them as small as possible. At the same time, we need to update resources faster and more frequently so that dependent services work properly.

### CODE REVIEW

File: InventoryBuilder.py [hunk 1]
```diff
@@ -134,44 +161,65 @@ def query_inventory_permanent(self):
...
+    def create_resource_objects(self, target, token):
+        vrops = Vrops()
+
+        name, uuid = Vrops.get_adapter(target, token)
+        if not name:
+            return False
+        logger.debug(f'Collecting vcenter: {name}')
+
+        datacenter = Vrops.get_resources(vrops, target, token, [uuid], resourcekinds=["Datacenter"])
+        vccluster = Vrops.get_resources(vrops, target, token, [dc.get('uuid') for dc in datacenter],
```
Comment (viennaa) [inline InventoryBuilder.py:192]: shouldn't we go with cluster here instead?

**Review by viennaa — COMMENTED** [review]
I really like the approach, going to have a massive speed and api call impact. Just few comments.

**Comment by viennaa** [comment]
Our tests are in an endless loop now.

File: InventoryBuilder.py [hunk 2]
```diff
@@ -134,44 +161,65 @@ def query_inventory_permanent(self):
...
+                                        resourcekinds=["ClusterComputeResource"])
+        hosts = Vrops.get_resources(vrops, target, token, [cl.get('uuid') for cl in vccluster],
+                                    resourcekinds=["HostSystem"])
+        vms_and_ds = Vrops.get_resources(vrops, target, token, [hs.get('uuid') for hs in hosts],
+                                         resourcekinds=["Datastore", "VirtualMachine"])
+        vms = [vm for vm in vms_and_ds if vm.get('resourcekind') == "VirtualMachine"]
+        dss = [ds for ds in vms_and_ds if ds.get('resourcekind') == "Datastore"]
+
+        vcenter = Vcenter(target, token, uuid, name)
+        vcenter.add_datacenter(datacenter[0])
```
[the remark this answers is not in the document]
Comment (richardtief) [reply InventoryBuilder.py]: We are sure. But I changed it to a loop anyway.

File: InventoryBuilder.py [hunk 3]
```diff
@@ -199,8 +247,8 @@ def get_datacenters(self):
                     'parent_vcenter_uuid': vcenter.uuid,
                     'parent_vcenter_name': vcenter.name,
                     'vcenter': vcenter.name,
-                    'target': dc.target,
-                    'token': dc.token,
+                    'target': vcenter.target,
+                    'token': vcenter.token,
```
[the remark this answers is not in the document]
Comment (richardtief) [reply InventoryBuilder.py:250]: To make API calls within the objects. No needed anymore.

File: InventoryBuilder.py (hunk 1 above)
Comment (richardtief) [reply InventoryBuilder.py]: Nope. We use the parent uuids to fetch the resources.

Comment (viennaa) [reply InventoryBuilder.py]: i meant variable naming `vcccluster`

Comment (richardtief) [reply InventoryBuilder.py]: suggestion pls

**Review by viennaa — APPROVED** [review]
Amazing finding, still. And shipping we do.

**Pull request merged**